In [ ]:
import pandas as pd
import numpy as np
import sklearn as sk
# import matplotlib.pyplot as plt
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.ensemble import AdaBoostClassifier
from google.colab import files
from geopy.geocoders import Nominatim
from sklearn import preprocessing
import itertools

# Get the Data from Kaggle

Credit to Kaggle user Filemon [1] for the utility code used here to get our dataset from Kaggle. It was copied and pasted with some minor alterations. Prof. Numanagic confirmed on Piazza that using this utility code is fine so long as we cite our source.

In [ ]:
files.upload()
! mkdir ~/.kaggle
! cp kaggle.json ~/.kaggle/
! chmod 600 ~/.kaggle/kaggle.json
! kaggle datasets list
! kaggle datasets download -d ruchi798/housing-prices-in-metropolitan-areas-of-india -p /content/sample_data/ --unzip

Saving kaggle.json to kaggle.json
mkdir: cannot create directory ‘/root/.kaggle’: File exists
ref                                                         title                                              size  lastUpdated          downloadCount  
----------------------------------------------------------  ------------------------------------------------  -----  -------------------  -------------  
gpreda/reddit-vaccine-myths                                 Reddit Vaccine Myths                              233KB  2021-08-12 07:39:36          10578  
crowww/a-large-scale-fish-dataset                           A Large Scale Fish Dataset                          3GB  2021-04-28 17:03:01           6463  
imsparsh/musicnet-dataset                                   MusicNet Dataset                                   22GB  2021-02-18 14:12:19           2421  
dhruvildave/wikibooks-dataset                               Wikibooks Dataset                                   2GB  2021-07-03 18:37:20

In [ ]:
cities = ['Bangalore', 'Chennai', 'Delhi', 'Hyderabad', 'Kolkata', 'Mumbai']
df1 = pd.read_csv("/content/sample_data/Bangalore.csv")
df2 = pd.read_csv("/content/sample_data/Chennai.csv")
df3 = pd.read_csv("/content/sample_data/Delhi.csv")
df4 = pd.read_csv("/content/sample_data/Hyderabad.csv")
df5 = pd.read_csv("/content/sample_data/Kolkata.csv")
df6 = pd.read_csv("/content/sample_data/Mumbai.csv")
df1.head()

Price  Area  ... Wardrobe  Refrigerator
0  30000000  3340  ...        0             0
1   7888000  1045  ...        0             0
2   4866000  1179  ...        0             0
3   8358000  1675  ...        0             0
4   6845000  1670  ...        0             0

[5 rows x 40 columns]

In [ ]:
dfs = [df1, df2, df3, df4, df5, df6]
for (i,k) in zip(dfs,cities):
  print(f"{k:<10} shape: {i.shape}")

Bangalore  shape: (6207, 40)
Chennai    shape: (5014, 40)
Delhi      shape: (4998, 40)
Hyderabad  shape: (2518, 40)
Kolkata    shape: (6507, 40)
Mumbai     shape: (7719, 40)


In [ ]:
dfs[0].describe()

Price         Area  ...     Wardrobe  Refrigerator
count  6.207000e+03  6207.000000  ...  6207.000000   6207.000000
mean   1.058510e+07  1526.094248  ...     6.171097      6.210569
std    1.410943e+07   764.845609  ...     4.178546      4.124430
min    2.000000e+06   415.000000  ...     0.000000      0.000000
25%    5.000000e+06  1110.000000  ...     0.000000      0.000000
50%    7.368000e+06  1340.000000  ...     9.000000      9.000000
75%    1.070000e+07  1662.500000  ...     9.000000      9.000000
max    3.000000e+08  9900.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

In [ ]:
dfs[1].describe()

Price         Area  ...     Wardrobe  Refrigerator
count  5.014000e+03  5014.000000  ...  5014.000000   5014.000000
mean   8.731264e+06  1232.992022  ...     4.991823      5.009174
std    1.034261e+07   641.841841  ...     4.473489      4.455995
min    2.000000e+06   362.000000  ...     0.000000      0.000000
25%    4.095000e+06   865.000000  ...     0.000000      0.000000
50%    5.782000e+06  1067.000000  ...     9.000000      9.000000
75%    8.675000e+06  1438.750000  ...     9.000000      9.000000
max    1.900000e+08  7240.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

In [ ]:
dfs[2].describe()

Price          Area  ...     Wardrobe  Refrigerator
count  4.998000e+03   4998.000000  ...  4998.000000   4998.000000
mean   1.588638e+07   1277.343737  ...     5.394958      5.411565
std    4.612894e+07    915.217800  ...     4.410549      4.392045
min    2.000000e+06    200.000000  ...     0.000000      0.000000
25%    2.900000e+06    800.000000  ...     0.000000      0.000000
50%    5.500000e+06   1000.000000  ...     9.000000      9.000000
75%    1.580000e+07   1650.000000  ...     9.000000      9.000000
max    8.546000e+08  16000.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

In [ ]:
dfs[3].describe()

Price         Area  ...     Wardrobe  Refrigerator
count  2.518000e+03  2518.000000  ...  2518.000000   2518.000000
mean   9.818380e+06  1644.571088  ...     0.341144      0.324861
std    8.777113e+06   746.256956  ...     1.621023      1.619344
min    2.000000e+06   500.000000  ...     0.000000      0.000000
25%    4.760000e+06  1160.000000  ...     0.000000      0.000000
50%    7.754000e+06  1500.000000  ...     0.000000      0.000000
75%    1.090000e+07  1829.000000  ...     0.000000      0.000000
max    1.650000e+08  9400.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

In [ ]:
dfs[4].describe()

Price          Area  ...     Wardrobe  Refrigerator
count  6.507000e+03   6507.000000  ...  6507.000000   6507.000000
mean   8.508570e+06   1344.166897  ...     8.896266      8.897034
std    1.465939e+07    807.698043  ...     0.960724      0.953984
min    2.000000e+06    350.000000  ...     0.000000      0.000000
25%    3.300000e+06    920.000000  ...     9.000000      9.000000
50%    4.940000e+06   1144.000000  ...     9.000000      9.000000
75%    8.000000e+06   1454.000000  ...     9.000000      9.000000
max    5.040000e+08  10000.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

In [ ]:
dfs[5].describe()

Price         Area  ...     Wardrobe  Refrigerator
count  7.719000e+03  7719.000000  ...  7719.000000   7719.000000
mean   1.506165e+07   998.409250  ...     7.383988      7.374789
std    2.052100e+07   550.967809  ...     3.438345      3.456702
min    2.000000e+06   200.000000  ...     0.000000      0.000000
25%    5.300000e+06   650.000000  ...     9.000000      9.000000
50%    9.500000e+06   900.000000  ...     9.000000      9.000000
75%    1.700000e+07  1177.000000  ...     9.000000      9.000000
max    4.200000e+08  8511.000000  ...     9.000000      9.000000

[8 rows x 39 columns]

Change the name for Children'splayarea to KidsPlayArea

In [ ]:
dfs = [df1, df2, df3, df4, df5, df6]
for i, k in enumerate(dfs):
  dfs[i] = k.rename(columns={"Children'splayarea":"KidsPlayArea"})

Credit to Kaggle user Pradyut Lohani [2] for identifiying and replacing "9" values with NaNs. "9" was used in the original dataset to mention absence of information; by replacing 9s with NaNs, we can use data that has some missing values without it throwing off our classifier.

In [ ]:
for i, k in enumerate(dfs):
  k.replace(9,np.nan, inplace=True)

We select the columns which have missing data to either replace them with simple imputation or delete samples with missing data

In [ ]:
col_names = dfs[0].columns
col_full_vals = ['Price', 'Area', 'Location', 'No. of Bedrooms', 'Resale']
col_not_full_vals = [col_names for col_names in col_names if col_names not in col_full_vals]

**Uncomment the following it to get datasets with simple imputation**

In [ ]:
for j in dfs:
  for i in col_not_full_vals:
    j[i].fillna(j[i].mean(), inplace = True)

In [ ]:
dfs[0].isna().sum() # count of how many properties are missing each type of data for Bangalore

Price                  0
Area                   0
Location               0
No. of Bedrooms        0
Resale                 0
MaintenanceStaff       0
Gymnasium              0
SwimmingPool           0
LandscapedGardens      0
JoggingTrack           0
RainWaterHarvesting    0
IndoorGames            0
ShoppingMall           0
Intercom               0
SportsFacility         0
ATM                    0
ClubHouse              0
School                 0
24X7Security           0
PowerBackup            0
CarParking             0
StaffQuarter           0
Cafeteria              0
MultipurposeRoom       0
Hospital               0
WashingMachine         0
Gasconnection          0
AC                     0
Wifi                   0
KidsPlayArea           0
LiftAvailable          0
BED                    0
VaastuCompliant        0
Microwave              0
GolfCourse             0
TV                     0
DiningTable            0
Sofa                   0
Wardrobe               0
Refrigerator           0


In [ ]:
            # def rename_location(dfs):
            #   for p in dfs:
            #     a = p['Location'].tolist()
            #     a_sorted = sorted(list(set(a)), key=str.lower)
            #     b = {}
            #     t = 0
            #     for i in a_sorted:
            #       b[i] = t
            #       t = t+1
            #     p['LocationRename'] = p['Location'].replace(b)

In [ ]:
            # dfs = [df1, df2, df3, df4, df5, df6]
            # rename_location(dfs)

In [ ]:
            # df1 = pd.get_dummies(df1)
            # df2 = pd.get_dummies(df2)
            # df3 = pd.get_dummies(df3)
            # df4 = pd.get_dummies(df4)
            # df5 = pd.get_dummies(df5)
            # df6 = pd.get_dummies(df6)

In [ ]:
for (i,k) in zip(dfs,cities):
  print(f"{k:<10} shape: {i.shape}")

Bangalore  shape: (6207, 40)
Chennai    shape: (5014, 40)
Delhi      shape: (4998, 40)
Hyderabad  shape: (2518, 40)
Kolkata    shape: (6507, 40)
Mumbai     shape: (7719, 40)


In [ ]:
geolocator = Nominatim(user_agent="housing")

def findLongLat(local):
  try:
    cordinates=geolocator.geocode(local)
    return cordinates.latitude, cordinates.longitude
  except:
    return "NA", "NA"

In [ ]:
def locationFindAndReplace(dfs):
  for j, (i, k) in enumerate(zip(dfs, cities)):
    df_selection = i
    city_selection = k
    lolk= df_selection['Location'].unique()
    lolp = [ str(t) +', '+str(city_selection) +', India' for t in lolk]
    locationLongLat={}
    print(f'{i.shape}, {city_selection}')
    for p, place in zip(lolk, lolp):
      locationLongLat[p]=findLongLat(place)
      print(place)
    dfs[j]['Latitude'], dfs[j]['Longitude'] = zip(*dfs[j]['Location'].map(locationLongLat))
    dfs[j].replace('NA', np.nan, inplace=True)
    dfs[j] = dfs[j].drop(columns='Location')
  return dfs

In [ ]:
dfs = locationFindAndReplace(dfs)

(6207, 40), Bangalore
JP Nagar Phase 1, Bangalore, India
Dasarahalli on Tumkur Road, Bangalore, India
Kannur on Thanisandra Main Road, Bangalore, India
Doddanekundi, Bangalore, India
Kengeri, Bangalore, India
Horamavu, Bangalore, India
Thanisandra, Bangalore, India
Ramamurthy Nagar, Bangalore, India
Whitefield Hope Farm Junction, Bangalore, India
Electronic City Phase 1, Bangalore, India
Yelahanka, Bangalore, India
Anjanapura, Bangalore, India
Jalahalli, Bangalore, India
Kasavanahalli, Bangalore, India
Bommasandra, Bangalore, India
Bellandur, Bangalore, India
RR Nagar, Bangalore, India
Begur, Bangalore, India
Hosa Road, Bangalore, India
Sahakar Nagar, Bangalore, India
Kadugodi, Bangalore, India
Jakkur, Bangalore, India
Jigani, Bangalore, India
Krishnarajapura, Bangalore, India
Brookefield, Bangalore, India
Banashankari, Bangalore, India
Nelamangala, Bangalore, India
Attibele, Bangalore, India
Banaswadi, Bangalore, India
Kodigehalli, Bangalore, India
ITPL, Bangalore, India
Uttarahalli H

In [ ]:
# dfs[0]['Latitude'], dfs[0]['Longitude'] = zip(*dfs[0]['Location'].map(locationLongLat))
# df1['Latitude'],df1['Longitude']=zip(*df1['Location'].map(locationLongLat))
# df2['Latitude'],df2['Longitude']=zip(*df2['Location'].map(locationLongLat))
# df3['Latitude'],df3['Longitude']=zip(*df3['Location'].map(locationLongLat))
# df4['Latitude'],df4['Longitude']=zip(*df4['Location'].map(locationLongLat))
# df5['Latitude'],df5['Longitude']=zip(*df5['Location'].map(locationLongLat))
# df6['Latitude'],df6['Longitude']=zip(*df6['Location'].map(locationLongLat))
# # convertLocation(df4)

In [ ]:
# for i in dfs:
#   i.replace('NA', np.nan, inplace=True)
# df1.replace('NA', np.nan, inplace=True)
# df2.replace('NA', np.nan, inplace=True)
# df3.replace('NA', np.nan, inplace=True)
# df4.replace('NA', np.nan, inplace=True)
# df5.replace('NA', np.nan, inplace=True)
# df6.replace('NA', np.nan, inplace=True)

In [ ]:
for (i,k) in zip(dfs,cities):
  print(f"{k:<10} shape: {i.shape}")

Bangalore  shape: (6207, 41)
Chennai    shape: (5014, 41)
Delhi      shape: (4998, 41)
Hyderabad  shape: (2518, 41)
Kolkata    shape: (6507, 41)
Mumbai     shape: (7719, 41)


Total sample with missing Values; after removing missing location values

*   Bangalore: 4256 of 6207: 1951: 1530
*   Chennai: 2781 of 5014: 2233: 1917
*   Delhi: 2996 of 4998: 2002: 1949
*   Hyderabad: 84 of 2518: 2434: 2174
*   Kolkata: 6507 of 6432: 75: 65
*   Mumbai: 6321 of 7719: 1398: 1114

Imputation:
* Bangalore  shape: (4914, 41)
* Chennai    shape: (4287, 41)
* Delhi      shape: (4747, 41)
* Hyderabad  shape: (2251, 41)
* Kolkata    shape: (5553, 41)
* Mumbai     shape: (6806, 41)

In [ ]:
for i in dfs:
  i.dropna(inplace=True)
# df1.dropna(inplace= True)
# df2.dropna(inplace= True)
# df3.dropna(inplace= True)
# df4.dropna(inplace= True)
# df5.dropna(inplace= True)
# df6.dropna(inplace= True)

In [ ]:
for (i,k) in zip(dfs,cities):
  print(f"{k:<10} shape: {i.shape}")

Bangalore  shape: (4914, 41)
Chennai    shape: (4287, 41)
Delhi      shape: (4747, 41)
Hyderabad  shape: (2251, 41)
Kolkata    shape: (5553, 41)
Mumbai     shape: (6806, 41)


In [ ]:
dfs[0].isna().sum()

Price                  0
Area                   0
No. of Bedrooms        0
Resale                 0
MaintenanceStaff       0
Gymnasium              0
SwimmingPool           0
LandscapedGardens      0
JoggingTrack           0
RainWaterHarvesting    0
IndoorGames            0
ShoppingMall           0
Intercom               0
SportsFacility         0
ATM                    0
ClubHouse              0
School                 0
24X7Security           0
PowerBackup            0
CarParking             0
StaffQuarter           0
Cafeteria              0
MultipurposeRoom       0
Hospital               0
WashingMachine         0
Gasconnection          0
AC                     0
Wifi                   0
KidsPlayArea           0
LiftAvailable          0
BED                    0
VaastuCompliant        0
Microwave              0
GolfCourse             0
TV                     0
DiningTable            0
Sofa                   0
Wardrobe               0
Refrigerator           0
Latitude               0


In [ ]:
# df1 = df1.drop(columns='Location')
# df2 = df2.drop(columns='Location')
# df3 = df3.drop(columns='Location')
# df4 = df4.drop(columns='Location')
# df5 = df5.drop(columns='Location')
# df6 = df6.drop(columns='Location')

In [ ]:
# dfs = [df1, df2, df3, df4, df5, df6]
for i, k in enumerate(dfs):
  scalar = preprocessing.MinMaxScaler()
  x = k.values
  x_scaled = scalar.fit_transform(x)
  dfs[i] = pd.DataFrame(x_scaled, columns=k.columns)

In [ ]:
dfs[0].head()

Price      Area  No. of Bedrooms  ...  Refrigerator  Latitude  Longitude
0  0.093960  0.308382         0.500000  ...           0.0  0.034778   0.293278
1  0.021336  0.132841         0.333333  ...           0.0  0.034180   0.303187
2  0.016258  0.132314         0.333333  ...           0.0  0.030072   0.283146
3  0.016097  0.084871         0.166667  ...           0.0  0.038469   0.299765
4  0.060403  0.220032         0.500000  ...           0.0  0.040565   0.297294

[5 rows x 41 columns]

In [ ]:
dfs[0].to_csv('BangaloreDataMunging.csv')
dfs[1].to_csv('ChennaiDataMunging.csv')
dfs[2].to_csv('DelhiDataMunging.csv')
dfs[3].to_csv('HyderabadDataMunging.csv')
dfs[4].to_csv('KolkataDataMunging.csv')
dfs[5].to_csv('MumbaiDataMunging.csv')

# Citations

[1] https://www.kaggle.com/general/74235

[2] https://www.kaggle.com/ruchi798/housing-prices-eda-and-prediction
